source/extensions/access_loggers/grpc/grpc_access_log_impl.cc

```diff
@@ -65,23 +64,23 @@ bool GrpcAccessLoggerImpl::isCriticalMessageEmpty() {
 }
 
 void GrpcAccessLoggerImpl::flushCriticalMessage() {
-  if (critical_client_ == nullptr || isCriticalMessageEmpty()) {
+  if (isCriticalMessageEmpty()) {
     return;
   }
-  if (!critical_client_->isStreamStarted()) {
+  if (!critical_logger_->shouldSetLogIdentifier()) {
     initCriticalMessage();
   }
```

Comment:
I think this is probably fine but I think a nicer pattern would be what I mentioned last time, where you provide a base RequestMessage that gets merged into the first message on the stream. That way the caller doesn't have to be concerned about having to modify the payload for the first message on the stream. WDYT?

Reply:
refactored. But I think if we have any grpc client which won't do set identifier like ot. we should implement virtual function to init message on logger.